In [2]:
import tensorflow as tf
import keras 


'''CONSTANT'''
# Here you just put the values/list/matrix/tensor in the first entry of tf.constant and it just does one calculation


node1 = tf.constant(3.0,'float32')
node2 = tf.constant(4.0)

print(node1,node2)   #this only prints the module details

# Initialize Session and run `result`
sess = tf.Session()
print(sess.run([node1,node2]))
sess.close()



Tensor("Const:0", shape=(), dtype=float32) Tensor("Const_1:0", shape=(), dtype=float32)
[3.0, 4.0]


Using TensorFlow backend.


In [3]:
'''This is a better way to run the code as it automatically the session after output is satisfied'''

with tf.Session() as sess:
  output = sess.run([node1,node2])
  print(output)
        

[3.0, 4.0]


In [5]:
'''Example 2'''

a = tf.constant(5.0)
b = tf.constant(6.0)
c = a*b

with tf.Session() as sess:
    print(sess.run(c))
    
File_Writer = tf.summary.FileWriter('graph',sess.graph)    

30.0


In [27]:
'''PLACE HOLDER'''

#Place Holder is a follow on from constant which is kind of like a variable 
# BUT you won't be able to change numbers after each iteration

a = tf.placeholder('float32')
b = tf.placeholder('float32')

add_noder = a+b 

with tf.Session() as sess:
    print(sess.run(add_noder,{a:[1,2],b:[2,3]}))
    
    
#As you can see, I am only running the code for a=[1,2] and b=[2,3] but it can not iterate 

[3. 5.]


In [34]:
'''Variable'''

# if i want my data to become trainable, I need parameters which change after every iteration
# you use variables to HOLD and UPDATE the parameters (each iteration)

import tensorflow as tf

#Model Parameters (i.e. we need to change the value of these after each iteration thus assign it as a variable not placeholder)
W = tf.Variable([0.3],'float32')
b = tf.Variable([-0.3],'float32')

#inputs and outputs 
x = tf.placeholder('float32')    #input
y = tf.placeholder('float32')    #desired output (i.e. training data)    

y_hat = W*x+b    #approximating with weights 
squared_delta = tf.square(y-y_hat)
loss = tf.reduce_sum(squared_delta)

#initialising the Variables which needs to be done each time
init = tf.global_variables_initializer()
sess = tf.Session()
sess.run(init)

print(sess.run(loss,{x:[1,2,3,4],y:[0,-1,-2,-3]}))

print('The loss I get: {}'.format(sess.run(loss,{x:[1,2,3,4],y:[0,-1,-2,-3]})))

23.66
The loss I get: 23.65999984741211


Here I will be describing the above code: 

$W$ = weights which is a variable and I have taken to be just $0.3$ (but you have to express it as $[0.3]$

$b$ = bias which is a variable and I have taken to be just $-0.3$ (but you have to express it as $[-0.3]$

$x$ and $y$ are inputs and *desired* output respectively and I have taken them as placeholder because I do not have to change them in each iteration 

$$y_{hat} = Wx+b \quad \text{and} \quad loss = \Sigma_i (y^{i}-y^{i}_{hat})^2$$

where $\delta^2 = (y^{i}-y^{i}_{hat})^2$

**Everytime you need to initialize the global variables so some steps are mandatory** 

Finally I chose $x = [1,2,3,4]$ and $y = [0,-1,-2,-3]$ as my numbers for the placeholder to get the desired result of 23.66

### Steepest Gradient Descent

let $\quad J(W_i) = \frac{1}{2}\Sigma_i (y^{i}-y^{i}_{hat})^2 = \Sigma_i (y^{i}-(W_ix_i+b))^2$, 
$$\Delta W_i = -\eta \frac{\partial J}{\partial W_i} \implies -\eta \Sigma_i (y^{i}-y^{i}_{hat})(-x_i)$$
$$ W = W + \Delta W $$ 




In [48]:
'''Implementing Steepest Gradient Descent Algorithm'''

W = tf.Variable([0.3],'float32')
b = tf.Variable([-0.3],'float32')

#inputs and outputs 
x = tf.placeholder('float32')    #input
y = tf.placeholder('float32')    #desired output (i.e. training data)    

y_hat = W*x+b    #approximating with weights 

#loss
squared_delta = tf.square(y-y_hat)
loss = tf.reduce_sum(squared_delta)

#optimize
optimizer = tf.train.GradientDescentOptimizer(0.01)    #this represents my stepsize (eta = 0.01) which I keep constant 
train = optimizer.minimize(loss)

#initialising the Variables which needs to be done each time
init = tf.global_variables_initializer()
sess = tf.Session()
sess.run(init)

for i in range(1000):
    sess.run(train,{x:[1,2,3,4],y:[0,-1,-2,-3]})

print(sess.run([W,b]))
    
#print(sess.run(loss,{x:[1,2,3,4],y:[0,-1,-2,-3]}))



[array([-0.9999969], dtype=float32), array([0.9999908], dtype=float32)]
